File: LeetCode/Study Plan/Practice Problems/Medium/Combination Sum/Go/combination_sum.go
package main

import "fmt"

func combination_sum(arr []int, target int) [][]int {
	res := [][]int{}
	_combiantion_sum(arr, target, []int{}, 0, &res)
	return res
}

func _combiantion_sum(arr []int, target int, set []int, ind int, res *[][]int) {
	if ind >= len(arr) {
		if target == 0 {
			*res = append(*res, append([]int{}, set...))
		}
		return
	}

	if arr[ind] <= target {
		temp := append(set, arr[ind])
		_combiantion_sum(arr, target-arr[ind], temp, ind, res)
		temp = temp[:len(temp)-1]
	}
	_combiantion_sum(arr, target, set, ind+1, res)

}

func main() {
	fmt.Println(combination_sum([]int{2, 3, 6, 7}, 7))
}

File: Algorithms/Coding Interview Patterns/Backtracking/combination_sum.cc
#include <iostream>
#include <vector>
#include <deque>

using namespace std;

void dfs(vector<int> arr, int sum, vector<vector<int>>& res, int curr_sum, deque<int> dq, int ind) {
    if (curr_sum == sum) {
        res.push_back({dq.begin(), dq.end()});
        return;
    }
    if (curr_sum > sum) return;

    for (int i = ind; i < arr.size(); i++) {
        dq.push_back(arr[i]);
        dfs(arr, sum, res, curr_sum + arr[i], dq, i);
        dq.pop_back();
    }
}


vector<vector<int>> combination_sum(vector<int> arr, int sum) {
    vector<vector<int>> res;
    dfs(arr, sum, res, 0, {}, 0);
    return res;
}

int main(int argc, char** argv) {
    vector<int> arr = {1, 2, 3};
    vector<vector<int>> res = combination_sum(arr, 4);
    for (auto vec : res) {
        for (auto val : vec) {
            cout << val << " ";
        }
        cout << "\n";
    }
    return 0;
}
